### PR TITLE
Command buffer and queue interfaces

### PR DESCRIPTION
**Pull request by Crisspl**
## Description
Discussion in https://github.com/Devsh-Graphics-Programming/Nabla/issues/3

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
Make sure to merge master after I merge armageddon

File: include/irr/video/IGPUCommandBuffer.h [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+class IGPUQueue;
+
+class IGPUCommandBuffer : public core::IReferenceCounted
+{
+    friend class IGPUQueue;
+
+public:
+    enum E_RESET_FLAGS : uint32_t
+    {
+        ERF_RELEASE_RESOURCES_BIT = 0x01
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:115]: document this flag

File: include/irr/video/IGPUCommandBuffer.h [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    {
+
+    }
+
+    virtual E_LEVEL getLevel() const = 0;
+
+    uint32_t getQueueFamilyIndex() const { return m_familyIndex; }
+
+    //! `_flags` takes bits from E_RESET_FLAGS
+    virtual void reset(uint32_t _flags)
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:150]: then dont make the flag type a `uint32_t`

File: include/irr/video/IGPUCommandBuffer.h [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    {
+        assert(m_state!=ES_PENDING);
+        m_state = ES_INITIAL;
+    }
+
+    virtual void end()
+    {
+        assert(m_state!=ES_PENDING);
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:158]: wouldn't you need to assert that the state must have been `ES_RECORDING`

File: include/irr/video/IGPUCommandBuffer.h [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    //virtual void clearAttachments(uint32_t attachmentCount, const VkClearAttachment* pAttachments, uint32_t rectCount, const VkClearRect* pRects) = 0;
+    virtual void fillBuffer(IGPUBuffer* dstBuffer, size_t dstOffset, size_t size, uint32_t data) = 0;
+    virtual void updateBuffer(IGPUBuffer* dstBuffer, size_t dstOffset, size_t dataSize, const void* pData) = 0;
+
+protected:
+    //! `_flags` takes bits from E_USAGE
+    virtual void begin(uint32_t _flags)
+    {
+        assert(m_state!=ES_PENDING);
+        assert(m_state!=ES_RECORDING);
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:243]: I think calling begin on a cmd buffer that is already `EXECUTABLE` is an error too


Doesn't a cmd buffer need to be in `INITIAL` state only, to begin recording?

File: include/irr/video/IGPUCommandBuffer.h [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        m_state = ES_EXECUTABLE;
+    }
+
+    virtual void bindIndexBuffer(IGPUBuffer* buffer, size_t offset, asset::E_INDEX_TYPE indexType) = 0;
+    virtual void draw(uint32_t vertexCount, uint32_t instanceCount, uint32_t firstVertex, uint32_t firstInstance) = 0;
+    virtual void drawIndexed(uint32_t indexCount, uint32_t instanceCount, uint32_t firstIndex, int32_t vertexOffset, uint32_t firstInstance) = 0;
+    virtual void drawIndirect(IGPUBuffer* buffer, size_t offset, uint32_t drawCount, uint32_t stride) = 0;
+    virtual void drawIndexedIndirect(IGPUBuffer* buffer, size_t offset, uint32_t drawCount, uint32_t stride) = 0;
+
+    //virtual void setViewport(uint32_t firstViewport, uint32_t viewportCount, const VkViewport* pViewports) = 0;
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:168]: what's the issue here?

File: include/irr/video/IGPUCommandBuffer.h [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+    virtual void setLineWidth(float lineWidth) = 0;
+    virtual void setDepthBias(float depthBiasConstantFactor, float depthBiasClamp, float depthBiasSlopeFactor) = 0;
+
+    virtual void setBlendConstants(const float blendConstants[4]) = 0;
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:173]: aren't they part of the renderpass?

File: include/irr/video/IGPUCommandBuffer.h [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    virtual void bindIndexBuffer(IGPUBuffer* buffer, size_t offset, asset::E_INDEX_TYPE indexType) = 0;
+    virtual void draw(uint32_t vertexCount, uint32_t instanceCount, uint32_t firstVertex, uint32_t firstInstance) = 0;
+    virtual void drawIndexed(uint32_t indexCount, uint32_t instanceCount, uint32_t firstIndex, int32_t vertexOffset, uint32_t firstInstance) = 0;
+    virtual void drawIndirect(IGPUBuffer* buffer, size_t offset, uint32_t drawCount, uint32_t stride) = 0;
+    virtual void drawIndexedIndirect(IGPUBuffer* buffer, size_t offset, uint32_t drawCount, uint32_t stride) = 0;
+
+    //virtual void setViewport(uint32_t firstViewport, uint32_t viewportCount, const VkViewport* pViewports) = 0;
+
+    virtual void setLineWidth(float lineWidth) = 0;
+    virtual void setDepthBias(float depthBiasConstantFactor, float depthBiasClamp, float depthBiasSlopeFactor) = 0;
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:171]: these can be baked into pipeline or be dynamic state correct?

File: include/irr/video/IGPUCommandBuffer.h [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+    virtual void setBlendConstants(const float blendConstants[4]) = 0;
+
+    virtual void copyBuffer(IGPUBuffer* srcBuffer, IGPUBuffer* dstBuffer, uint32_t regionCount, const VkBufferCopy* pRegions) = 0;
+    //TODO theres no E_LAYOUT enum!
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUCommandBuffer.h:176]: time to make one

File: include/irr/video/IGPUQueue.h [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            const auto* cmdbuf = _submit.commandBuffers[i];
+            submit(cmdbuf);
+        }
+    }
+    void submit_epilogue(IDriverFence* _fence)
+    {
+        if (_fence)
+            _fence->waitCPU(9999999999ull);
+    }
```
Comment (devshgraphicsprogramming) [inline include/irr/video/IGPUQueue.h:65]: what is this for?

Comment (Crisspl) [reply include/irr/video/IGPUQueue.h:65]: dont remember, look at usage

Comment (Crisspl) [reply include/irr/video/IGPUQueue.h:65]: maybe i've messed up something

Comment (devshgraphicsprogramming) [reply include/irr/video/IGPUQueue.h:65]: I haven't seen any ;)

Comment (Crisspl) [reply include/irr/video/IGPUQueue.h:65]: i thin it was supposed to be fence in this call https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkQueueSubmit.html but it should be signal not wait

Comment (devshgraphicsprogramming) [reply include/irr/video/IGPUQueue.h:65]: this sort of stuff do only in OpenGL backend... furthermore you should only use/signal a fence if you actually need to know when/if the command buffer submissions have been performed, and this is taken care of by specifying events or semaphores in the submit info.

**Pull request merged**